includes/class-products.php

```diff
@@ -95,103 +107,180 @@ public static function is_active() {
 		return class_exists( 'WooCommerce' ) && class_exists( 'WC_Subscriptions_Product' ) && defined( 'NEWSPACK_LISTINGS_SELF_SERVE_ENABLED' ) && NEWSPACK_LISTINGS_SELF_SERVE_ENABLED;
 	}
 
+	/**
+	 * Verify that the products associated with this feature set are valid and published.
+	 *
+	 * @return boolean True if valid.
+	 */
+	public static function validate_products() {
+		$valid    = true;
+		$products = self::get_products();
+
+		// Ensure that the products are published and valid.
+		if ( false === $products || is_wp_error( $products ) ) {
+			return false;
+		}
+
+		// Ensure that all parent and child product keys exist.
+		foreach ( self::PRODUCT_META_KEYS as $meta_key ) {
+			if ( ! isset( $products[ $meta_key ] ) ) {
+				return false;
+			}
+		}
+
+		return true;
+	}
+
 	/**
 	 * Init subclasses and create base products.
 	 */
 	public static function setup() {
-		if ( self::is_active() ) {
-			self::create_products();
-
+		if ( self::is_active() && self::validate_products() ) {
 			new Products_Purchase();
 			new Products_Ui();
 			new Products_User();
 			new Products_Cron();
 		}
 	}
 
+	/**
+	 * Create or delete self-serve listing products.
+	 */
+	public static function perform_action() {
```

Comment:
Nit: the method name is not too descriptive ;)

Reply:
Updated in 866f9b7